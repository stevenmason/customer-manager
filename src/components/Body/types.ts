import { Customer } from "../../store/types";
import {
  setEditingCustomer,
  setFilter,
  addNewEditingCustomer,
  deleteCustomer
} from "../../store/actions";

export interface Props {
  customers: Customer[];
  loading: boolean;
  filter: string;
  setEditingCustomer: typeof setEditingCustomer;
  setFilter: typeof setFilter;
  addNewEditingCustomer: typeof addNewEditingCustomer;
  deleteCustomer: typeof deleteCustomer;
}
