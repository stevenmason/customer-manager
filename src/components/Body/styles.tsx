import styled from "styled-components";

const Container = styled.div`
  margin-left: auto;
  margin-right: auto;
  max-width: 980px;
  padding: 30px;
  color: #98a4a6;
`;

const Underline = styled.u`
  cursor: pointer;
  float: right;
  margin: 20px 0;
`;

export { Underline, Container };
