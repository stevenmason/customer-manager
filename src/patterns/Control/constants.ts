enum SIZES {
  LARGE = "LARGE",
  SMALL = "SMALL"
}

export { SIZES };
