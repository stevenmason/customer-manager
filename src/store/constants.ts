enum ACTIONS {
  SET_CUSTOMERS = "SET_CUSTOMERS",
  SET_CUSTOMER = "SET_CUSTOMER",
  DELETE_CUSTOMER = "DELETE_CUSTOMER",
  SET_EDITING_CUSTOMER = "SET_EDITING_CUSTOMER",
  SET_EDITING_CUSTOMER_VALUE = "SET_EDITING_CUSTOMER_VALUE",
  SET_LOADING = "SET_LOADING",
  SET_FILTER = "SET_FILTER"
}

const DEFAULT_CUSTOMER = {
  id: "-1",
  firstName: "",
  lastName: "",
  dateOfBirth: ""
};

export { ACTIONS, DEFAULT_CUSTOMER };
