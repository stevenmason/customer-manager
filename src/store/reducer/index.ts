import { combineReducers } from "redux";

import customers from "./customers";
import editingCustomer from "./editingCustomer";
import filter from "./filter";
import loading from "./loading";

const rootReducer = combineReducers({
  customers,
  filter,
  editingCustomer,
  loading
});

export default rootReducer;
