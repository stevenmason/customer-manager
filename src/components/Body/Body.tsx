import React from "react";

import Toolbar from "./Toolbar";
import CustomerItem from "./CustomerItem";
import { Props } from "./types";
import * as SC from "./styles";
import List from "../../patterns/List";

const scrollToTop = () => window.scroll({ top: 0, behavior: "smooth" });

const Body: React.FC<Props> = ({
  customers,
  loading,
  setEditingCustomer,
  addNewEditingCustomer,
  filter,
  setFilter,
  deleteCustomer
}) => {
  console.log("render");
  return (
    <SC.Container>
      <Toolbar
        filter={filter}
        setFilter={setFilter}
        addNewEditingCustomer={addNewEditingCustomer}
      />
      <List loading={loading}>
        {customers.map(customer => (
          <CustomerItem
            key={customer.id}
            customer={customer}
            setEditingCustomer={setEditingCustomer}
            deleteCustomer={deleteCustomer}
          />
        ))}
      </List>
      <SC.Underline onClick={scrollToTop}>Back to top</SC.Underline>
    </SC.Container>
  );
};

export default Body;
