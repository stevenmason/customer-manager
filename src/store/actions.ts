import { uniqueId } from "lodash";

import { ACTIONS, DEFAULT_CUSTOMER } from "./constants";
import { Customer } from "./types";

const setCustomers = (customers: Customer[]) => ({
  type: ACTIONS.SET_CUSTOMERS as ACTIONS.SET_CUSTOMERS,
  customers
});

const setCustomer = (customer: Customer) => ({
  type: ACTIONS.SET_CUSTOMER as ACTIONS.SET_CUSTOMER,
  customer
});

const deleteCustomer = (id: string) => ({
  type: ACTIONS.DELETE_CUSTOMER as ACTIONS.DELETE_CUSTOMER,
  id
});

const setEditingCustomer = (editingCustomer: Customer | null) => ({
  type: ACTIONS.SET_EDITING_CUSTOMER as ACTIONS.SET_EDITING_CUSTOMER,
  editingCustomer
});

const addNewEditingCustomer = () => ({
  type: ACTIONS.SET_EDITING_CUSTOMER as ACTIONS.SET_EDITING_CUSTOMER,
  editingCustomer: {
    ...DEFAULT_CUSTOMER,
    id: uniqueId()
  }
});

const setEditingCustomerValue = (value: string, name: string) => ({
  type: ACTIONS.SET_EDITING_CUSTOMER_VALUE as ACTIONS.SET_EDITING_CUSTOMER_VALUE,
  name,
  value
});

const setLoading = (loading: boolean) => ({
  type: ACTIONS.SET_LOADING as ACTIONS.SET_LOADING,
  loading
});

const setFilter = (filter: string) => ({
  type: ACTIONS.SET_FILTER as ACTIONS.SET_FILTER,
  filter
});

export {
  setCustomers,
  setCustomer,
  setEditingCustomer,
  setEditingCustomerValue,
  setLoading,
  setFilter,
  addNewEditingCustomer,
  deleteCustomer
};
