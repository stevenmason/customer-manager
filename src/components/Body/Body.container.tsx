import { connect } from "react-redux";
import { bindActionCreators, Dispatch, Action } from "redux";

import {
  setEditingCustomer,
  setFilter,
  addNewEditingCustomer,
  deleteCustomer
} from "../../store/actions";
import { filteredCustomers, loading, filter } from "../../store/selectors";
import { State } from "../../store/types";
import Body from "./Body";

const mapStateToProps = (state: State) => ({
  customers: filteredCustomers(state),
  loading: loading(state),
  filter: filter(state)
});

const mapDispatchToProps = (dispatch: Dispatch<Action>) => ({
  ...bindActionCreators(
    {
      setEditingCustomer,
      setFilter,
      addNewEditingCustomer,
      deleteCustomer
    },
    dispatch
  )
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Body);
