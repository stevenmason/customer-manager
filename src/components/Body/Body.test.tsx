import React from "react";
import renderer from "react-test-renderer";

import Body from "./Body";
import {
  setEditingCustomer,
  deleteCustomer,
  setFilter,
  addNewEditingCustomer
} from "../../store/actions";
import { DEFAULT_CUSTOMER } from "../../store/constants";

const PROPS = {
  filter: "",
  customers: [DEFAULT_CUSTOMER],
  loading: false,
  setEditingCustomer,
  setFilter,
  addNewEditingCustomer,
  deleteCustomer
};

it("renders correctly", () => {
  const tree = renderer.create(<Body {...PROPS} />);
  expect(tree.toJSON()).toMatchSnapshot();
});
