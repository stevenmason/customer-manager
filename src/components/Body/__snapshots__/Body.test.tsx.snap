// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="sc-dnqmqq eqfUuj"
>
  <div
    className="sc-bdVaJa fzbFad"
  >
    <div
      className="sc-htpNat zrlvf"
      size="LARGE"
    >
      <label
        className="sc-bxivhb ctASUB"
      >
        Filter
      </label>
      <input
        className="sc-ifAKCX jOZwNh"
        name="filter"
        onChange={[Function]}
        value=""
      />
    </div>
    <button
      className="sc-bwzfXH ppMEU"
      onClick={[Function]}
    >
      Add Customer
    </button>
  </div>
  <div
    className="sc-gZMcBi pTNGW"
  >
    <span
      className="sc-gqjmRU jcjFtD"
    >
      Showing 1 Customers
    </span>
    <div
      className="sc-gzVnrw jnQJaE"
    >
      <div>
        <div
          className="sc-htpNat bZZHvT"
          size="SMALL"
        >
          <label
            className="sc-bxivhb ctASUB"
          >
            Name:
          </label>
          <span
            className="sc-EHOje kyfnqM"
          >
             
          </span>
        </div>
        <div
          className="sc-htpNat bZZHvT"
          size="SMALL"
        >
          <label
            className="sc-bxivhb ctASUB"
          >
            Date of Birth:
          </label>
          <span
            className="sc-EHOje kyfnqM"
          >
            Invalid Date
          </span>
        </div>
      </div>
      <span
        className="sc-bZQynM kzlsZH"
      >
        <div
          className="sc-htoDjs bYJfks"
          onClick={[Function]}
        >
          <svg
            viewBox="0 0 24 24"
          >
            <path
              d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34c-.39-.39-1.02-.39-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
            />
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
          </svg>
        </div>
        <div
          className="sc-htoDjs bYJfks"
          onClick={[Function]}
        >
          <svg
            viewBox="0 0 24 24"
          >
            <path
              d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
            />
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
          </svg>
        </div>
      </span>
    </div>
  </div>
  <u
    className="sc-iwsKbI kPFatP"
    onClick={[Function]}
  >
    Back to top
  </u>
</div>
`;
